server/sonar-server/src/main/java/org/sonar/server/es/StartupIndexer.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.es;
+
+import java.util.Set;
+
+/**
+ * This kind of indexers get initialized during web server startup.
+ */
+public interface StartupIndexer {
+
+  void initializeOnStartup();
```

Comment:
Note that it is risky to re-index all the types if a single one is empty. Imagine that a lot of projects are analyzed but none have test information (for example cobol or PL/SQL projects). The authorization types will be truncated and re-indexed at each startup. That can significantly slow down startup.

An option is to add the parameter `initializeOnStartup(IndexTypeId)` so that only the specified type is indexed. Method is called as many times as empty types.